content/posts/unpublished/building-reliable-and-scalable-applications-a-guide-to-12-factor-apps-with-java-and-spring-boot/post.md

```diff
@@ -0,0 +1,324 @@
+            steps {
+                // Deployment steps
+            }
+        }
+    }
+}
+```
+Automating build, release, and run processes is essential for efficiency. GitHub Actions provides an excellent platform for CI/CD. Below is an example workflow for building and deploying a Spring Boot application:
+
+```yaml
+name: CI/CD
+
+on:
+  push:
+    branches:
+      - main
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
+
+    steps:
+    - name: Checkout code
+      uses: actions/checkout@v2
+
+    - name: Set up JDK
+      uses: actions/setup-java@v2
+      with:
+        java-version: '11'
+
+    - name: Build with Maven
+      run: mvn -B package --file pom.xml
+
+  deploy:
+    needs: build
+    runs-on: ubuntu-latest
+
+    steps:
+    - name: Deploy to Kubernetes
+      uses: azure/k8s-deploy@v1
+      with:
+        azure-k8s: '<your-azure-kubernetes-service>'
+        images: 'your-docker-image:latest'
+        namespace: 'default'
+        secrets: '<your-kubernetes-secrets>'
+```
+
+### 6. Processes
+
+Treat applications as one or more stateless processes. Each request should be handled independently without relying on persistent state within the process. Spring Boot applications typically run as stateless services, making them horizontally scalable. Spring Boot applications are standalone Java processes. You can run them using the `java -jar` command:
+
+```bash
+java -jar your-application.jar
+```
+
+### 7. Port Binding
+Bind applications to a port using environment variables. Spring Boot applications typically listen on a port defined in the application properties or environment variables.
+
+Spring Boot allows configuring the port dynamically using environment variables. Below is an example configuration in `application.properties`:
+
+```properties
+# Server port
+server.port=${PORT:8080}
+```
+
+### 8. Concurrency
+
+Leverage process-based concurrency for scaling. Spring Boot applications can be easily deployed across multiple instances to handle increased traffic.
+
+Java provides powerful concurrency utilities. Spring Boot supports asynchronous processing. Here's an example of an asynchronous task in a service:
+
+```java
+import org.springframework.scheduling.annotation.Async;
+import org.springframework.stereotype.Service;
+
+@Service
+public class MyService {
+
+    @Async
+    public void asyncTask() {
+        // Async task implementation
+    }
+}
+```
+
+### 9. Disposability
+
+Design applications to be fast to start and stop. Spring Boot applications typically have quick startup times, making them suitable for containerization.
+
+Spring Boot applications can gracefully handle shutdown signals. You can perform cleanup tasks using `@PreDestroy` annotation:
+
+```java
+import javax.annotation.PreDestroy;
+import org.springframework.stereotype.Component;
+
+@Component
+public class MyComponent {
+
+    @PreDestroy
+    public void cleanup() {
+        // Cleanup tasks
+    }
+}
+```
+
+### 10. Dev/Prod Parity
```

Comment:
I guess spring profiles worth mentioning here

Reply:
Externalized configuration does cover various ways

Reply:
Indeed, but profiles are still used, even in combination with external configuration.
Also, is that make sense to provide an example of using externalized configuration?

Reply:
Writing this blog is challenging given the fact that there are lot of insights can be provided in each factors of 12 factor app. Its difficult to include every aspect and also not to intimidate the reader with lot of tech. What I am trying to educate is about 12 factor app, and there are some factos which can be evolved as separate blog posts as well. Lets keep it short and crisp. On the point and less than 7 mins read.

Reply:
My point if "This post explores best practices for crafting robust and scalable applications in the backend space using Java and Spring Boot.", then using environment variables, profiles and externalized configuration should have a sample of its applying. Things like k8s, terraform, docker images is broader topic, I guess it worth mentioning too, but without overwhelming it with details as deployment manifest for k8s.

Reply:
this part has been changed recently to make the content consistent.